src/test_runner.rs

```diff
@@ -157,17 +157,47 @@ pub enum TestCaseError {
     ///
     /// The string gives the location and context of the rejection, and
     /// should be suitable for formatting like `Foo did X at {whence}`.
-    Reject(String),
+    Reject(Rejection),
     /// The code under test failed the test.
     ///
     /// The string should indicate the location of the failure, but may
     /// generally be any string.
-    Fail(String),
+    Fail(Rejection),
 }
 
 /// Convenience for the type returned by test cases.
 pub type TestCaseResult = Result<(), TestCaseError>;
 
+impl TestCaseError {
+    /// Rejects the generated test input as invalid for this test case. This
+    /// does not count as a test failure (nor a success); rather, it simply
+    /// signals to generate a new input and try again.
+    ///
+    /// The string gives the location and context of the rejection, and
+    /// should be suitable for formatting like `Foo did X at {whence}`.
+    pub fn reject<R: Into<Rejection>>(reason: R) -> Self {
+        TestCaseError::Reject(reason.into())
+    }
+
+    /// The code under test failed the test.
+    ///
+    /// The string should indicate the location of the failure, but may
+    /// generally be any string.
+    pub fn fail<R: Into<Rejection>>(reason: R) -> Self {
+        TestCaseError::Fail(reason.into())
+    }
+}
+
+/// Short-hand for `Err(TestCaseError::reject(..))`.
+pub fn reject_case<R: Into<Rejection>>(reason: R) -> TestCaseResult {
+    Err(TestCaseError::reject(reason))
+}
+
+/// Short-hand for `Err(TestCaseError::fail(..))`.
+pub fn fail_case<R: Into<Rejection>>(reason: R) -> TestCaseResult {
```

Comment:
These two are largely redundant with `prop_assert!` and `prop_assume!`, since both will virtually always be nested inside a condition.

Reply:
That's fair =) Can we make it non-`pub` for now? It adds some semantic compression inside `proptest` itself at places.